types/aframe/index.d.ts

```diff
@@ -89,62 +90,46 @@ declare namespace AFrame {
 		tick(): void;
 	}
 
-	interface Component {
+	interface Component<T extends { [key: string]: any } = any, S extends System | undefined = System | undefined> {
 		attrName?: string;
-		data?: any;
+		data: T;
 		dependencies?: string[];
 		el: Entity;
 		id: string;
 		multiple?: boolean;
 		name: string;
-		schema: Schema;
-
-		init(data?: any): void;
-		pause(): void;
-		play(): void;
-		remove(): void;
-		tick?(time: number, timeDelta: number): void;
-		update(oldData: any): void;
-		updateSchema?(): void;
-
-		extendSchema(update: Schema): void;
-		flushToDOM(): void;
+		schema: Schema<this['data']>;
+		system: S;
+
+		init(this: this, data?: this['data']): void;
```

Comment:
Hmm. I'm not so sure about this. When I remove `this: this`, `this` is typed as `Partial<TestComponent>` on line 64 of `aframe-tests.ts`, so I get an error because `this.data` might be undefined. Is there a way around that?

Reply:
I see what's happening. `this: this` works because I was lazy and defined `ComponentDefinition = Partial<Component>`.  In reality, A-Frame is using the definition to do prototype class construction: https://github.com/aframevr/aframe/blob/7af01068e410e8027cafe04dd5b567170aa8933c/src/core/component.js#L461-L482

The `ComponentDefinition` becomes the basis of the prototype of a new class. By using `this: this`, this means Partial<Component>, which disallows any properties not explicitly named by `Component`; e.g. helper functions and extra properties.

In the short term, I would suggest that we remove `this: this` and force users to explicitly declare `this`. Like:

```typescript
const Component = AFRAME.registerComponent<TestComponent>('test-component', {
	defaultValue: 0,
	init(this: AFrame.Component | { defaultValue: number }) {
		this.data.num = this.defaultValue;
	}
});
```

In the long term, we should update our types to use prototypical inheritance to better represent A-Frame's factory-produced class. Something similar to this: https://github.com/dojo/typings/blob/1a1a17e26f834203b58275ea6a2a7b8503539e8c/dojo/1.11/_base.d.ts#L578-L580

Reply:
Well, I added the `Partial` bit, so I guess that was me being lazy, though I'm not sure what it should be instead. I looked through the dojo typings you referenced, but though I think I understand how they are working, and I think I understand how A-Frame is doing prototype class construction, I'm not sure how to apply the one to the other! I'm fine with leaving this for a future update, but if you have the time to sketch out the ideal approach I'd be happy to give it a shot.

In the meantime, I must be missing something, but I still don't see why `this: this` is a problem.

> By using `this: this`, this means `Partial<Component>`, which disallows any properties not explicitly named by `Component`; e.g. helper functions and extra properties.  

Where is it that `this: this` disallows custom component properties? Do you mean in, for example, line 64 of `aframe-tests.ts`?:
```
const Component = AFRAME.registerComponent<TestComponent>('test-component', {
.
.
.
	init() {
		this.data.num = 0;
	},
```
 Its type there is actually `TestComponent`, which seems to me to be exactly what we want. It feels like there should be a cleaner way, and maybe the prototypical inheritance rewrite is that, but as far as I can tell it's doing what we want, which is forcing the `this` inside functions written in the component definition to take the type of the resultant component and not the type of the (partial) definition.

Reply:
hah sorry about the lazy quip. The `Partial` is totally something I would've done myself. I should've checked the history first 😊 .

Because of the way we've defined `registerComponent`, `this` is `Partial<Component>`.  It creates problems were we cannot reference custom properties provided in the definition:

<img width="1156" alt="screen shot 2018-04-28 at 5 25 55 pm" src="https://user-images.githubusercontent.com/331431/39402008-4ca96dfe-4b09-11e8-8bd3-a9b64ca4dc6c.png">

I've cleaned up my aframe-typings repo a bit (tests build but don't run ATM) and added your updates to `src` and started working on a prototype construction here: https://github.com/devpaul/aframe-typings/commit/353d1167617ccf56df45d459a9632252fabbdf63

I'm not 100% sure I have it right as I think `this` is being typed to `any` within the definition. Take a look and let me know what you think

Reply:
If you'd like to leave this prototype construction for a future update, we could remove the `this: this` typings and merge what you have. It already does a great job updating the AFrame typings. It would be a huge plus to have these in the repo 👍